File: test/benchmark/beng/js/fasta.js
// BENG Benchmark: fasta (Node.js reference)
// Generate DNA sequences using LCG PRNG

const N = parseInt(process.argv[2] || "1000");
const LINE_WIDTH = 60;

const ALU = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAAGG";

const IUB = [
    ['a', 0.27], ['c', 0.12], ['g', 0.12], ['t', 0.27],
    ['B', 0.02], ['D', 0.02], ['H', 0.02], ['K', 0.02],
    ['M', 0.02], ['N', 0.02], ['R', 0.02], ['S', 0.02],
    ['V', 0.02], ['W', 0.02], ['Y', 0.02]
];

const HOMO = [
    ['a', 0.3029549426680], ['c', 0.1979883004921],
    ['g', 0.1975473066391], ['t', 0.3015094502008]
];

function makeCumulative(table) {
    let sum = 0;
    for (const entry of table) {
        sum += entry[1];
        entry[1] = sum;
    }
}

let seed = 42;
const IM = 139968, IA = 3877, IC = 29573;

function random(max) {
    seed = (seed * IA + IC) % IM;
    return max * seed / IM;
}

function repeatFasta(id, desc, src, n) {
    console.log(`>${id} ${desc}`);
    let k = 0;
    let line = '';
    for (let i = 0; i < n; i++) {
        line += src[k];
        k = (k + 1) % src.length;
        if (line.length === LINE_WIDTH) {
            console.log(line);
            line = '';
        }
    }
    if (line.length > 0) console.log(line);
}

function randomFasta(id, desc, table, n) {
    console.log(`>${id} ${desc}`);
    makeCumulative(table);
    let line = '';
    for (let i = 0; i < n; i++) {
        const r = random(1.0);
        for (const [ch, prob] of table) {
            if (r < prob) { line += ch; break; }
        }
        if (line.length === LINE_WIDTH) {
            console.log(line);
            line = '';
        }
    }
    if (line.length > 0) console.log(line);
}

repeatFasta("ONE", "Homo sapiens alu", ALU, N * 2);
randomFasta("TWO", "IUB ambiguity codes", IUB, N * 3);
randomFasta("THREE", "Homo sapiens frequency", HOMO, N * 5);

File: test/benchmark/beng/python/fasta.py
#!/usr/bin/env python3
# BENG Benchmark: fasta (Python)
# Generate DNA sequences using LCG PRNG
import sys
import time

N = int(sys.argv[1]) if len(sys.argv) > 1 else 1000
LINE_WIDTH = 60

ALU = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAAGG"

IUB = [
    ('a', 0.27), ('c', 0.12), ('g', 0.12), ('t', 0.27),
    ('B', 0.02), ('D', 0.02), ('H', 0.02), ('K', 0.02),
    ('M', 0.02), ('N', 0.02), ('R', 0.02), ('S', 0.02),
    ('V', 0.02), ('W', 0.02), ('Y', 0.02)
]

HOMO = [
    ('a', 0.3029549426680), ('c', 0.1979883004921),
    ('g', 0.1975473066391), ('t', 0.3015094502008)
]


def make_cumulative(table):
    s = 0.0
    result = []
    for ch, prob in table:
        s += prob
        result.append((ch, s))
    return result


IM, IA, IC = 139968, 3877, 29573
seed = 42


def rand_next(max_val):
    global seed
    seed = (seed * IA + IC) % IM
    return max_val * seed / IM


def repeat_fasta(name, desc, src, n):
    print(f">{name} {desc}")
    src_len = len(src)
    k = 0
    buf = []
    for _ in range(n):
        buf.append(src[k])
        k = (k + 1) % src_len
        if len(buf) == LINE_WIDTH:
            print(''.join(buf))
            buf = []
    if buf:
        print(''.join(buf))


def random_fasta(name, desc, table, n):
    print(f">{name} {desc}")
    cum_table = make_cumulative(table)
    buf = []
    for _ in range(n):
        r = rand_next(1.0)
        ch = cum_table[-1][0]
        for c, prob in cum_table:
            if r < prob:
                ch = c
                break
        buf.append(ch)
        if len(buf) == LINE_WIDTH:
            print(''.join(buf))
            buf = []
    if buf:
        print(''.join(buf))


t0 = time.perf_counter_ns()
repeat_fasta("ONE", "Homo sapiens alu", ALU, N * 2)
random_fasta("TWO", "IUB ambiguity codes", IUB, N * 3)
random_fasta("THREE", "Homo sapiens frequency", HOMO, N * 5)
t1 = time.perf_counter_ns()
print(f"__TIMING__:{(t1 - t0) / 1e6:.3f}")
